coresdk/src/coresdk/raspi_gpio.cpp

```
@@ -229,4 +229,152 @@ namespace splashkit_lib
+        }
+        else
+        {
+            sk_remote_gpio_set_mode(pi, bcmPin, mode);
+        }
+    }
+
+    pin_modes remote_raspi_get_mode(connection pi, pins pin)
+    {
+        int bcmPin = boardToBCM(pin);
+        if (bcmPin == -1)
+        {
+            cout << "Cant modify a HIGH Pin" << endl;
+        }
+        else if (bcmPin == -2)
+        {
+            cout << "Cant modify a Ground pin" << endl;
+        }
+        else
+        {
+            return static_cast<pin_modes>(sk_remote_gpio_get_mode(pi, bcmPin));
+        }
+        return GPIO_DEFAULT_MODE;
+    }
+
+    void remote_raspi_set_pull_up_down(connection pi, pins pin, pull_up_down pud)
+    {
+        int bcmPin = boardToBCM(pin);
+        if (bcmPin == -1)
+        {
+            cout << "Cant modify a HIGH Pin" << endl;
+        }
+        else if (bcmPin == -2)
+        {
+            cout << "Cant modify a Ground pin" << endl;
+        }
+        else
+        {
+            sk_remote_gpio_set_pull_up_down(pi, bcmPin, pud);
+        }
+    }
+
+    void remote_raspi_write(connection pi, pins pin, pin_values value)
+    {
+        int bcmPin = boardToBCM(pin);
+        if (bcmPin == -1)
+        {
+            cout << "Cant modify a HIGH Pin" << endl;
+        }
+        else if (bcmPin == -2)
+        {
+            cout << "Cant modify a Ground pin" << endl;
+        }
+        else
+        {
+            sk_remote_gpio_write(pi, bcmPin, value);
+        }
+    }
+
+    pin_values remote_raspi_read(connection pi, pins pin)
+    {
+        int bcmPin = boardToBCM(pin);
+        if (bcmPin == -1)
+        {
+            cout << "Cant modify a HIGH Pin" << endl;
+        }
+        else if (bcmPin == -2)
+        {
+            cout << "Cant modify a Ground pin" << endl;
+        }
+        else
+        {
+            return static_cast<pin_values>(sk_remote_gpio_read(pi, bcmPin));
+        }
+        return GPIO_DEFAULT_VALUE;
```

I know splashkit never does error handling but I could see this causing some confusion as the function returns a valid output. I guess this is more of a statement rather then something to fix.

---

Yeah, I'd like to change this to some kind of error code with an associated error string.  This matches how the local functions work, so I'd like to change this when I get around to changing the local functions.

---

I think expanding `pin_values` to have a `GPIO_ERROR_VALUE` and returning that could work well? I think that can be returned if a network issue occurs as well.

From what I can tell, the local functions handle the ground/power pins by outputting a default value of high/low (which I think should be adopted here), while errors are undefined behavior since the code just casts whatever `gpio_read` returns to a `pin_value` (including negative error codes😅). Errors really should return an error code I think haha, maybe adding better error handling to the local functions should also be a task.

Also just off topic, but I noticed the local `sk_gpio_read` doesn't return a value if `check_pi` fails, which is also undefined behavior - we should add fixing that as a task too!